persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedDataStore.java

```diff
@@ -1,81 +1,178 @@
-            prepared ->
-                new PersistenceBackedPreparedStatement(
-                    connection, parameters, new PreparedInfo(prepared), queryString));
+        .prepare(query.queryStringForPreparation(), parameters())
+        .thenApply(prepared -> query.withPreparedId(prepared.statementId));
   }
 
   @Override
-  public CompletableFuture<ResultSet> batch(List<String> queries) {
+  public CompletableFuture<ResultSet> execute(
+      BoundQuery query, UnaryOperator<Parameters> parametersModifier) {
     long queryStartNanos = System.nanoTime();
-    List<Statement> persistenceStatements =
-        queries.stream().map(SimpleStatement::new).collect(Collectors.toList());
-    return batch(persistenceStatements, BatchType.LOGGED, parameters, queryStartNanos);
+    Parameters executeParameters = parametersModifier.apply(parameters());
+    validateExecuteParameters(executeParameters);
+
+    CompletableFuture<ResultSet> future = new CompletableFuture<>();
+    if (query.bounded().query().preparedId().isPresent()) {
+      executeWithRetry(query, executeParameters, queryStartNanos, future);
+    } else if (options.alwaysPrepareQueries()) {
+      prepareAndRetry(query.bounded(), executeParameters, queryStartNanos, future);
+    } else {
+      doExecute(query, executeParameters, queryStartNanos, future, future::completeExceptionally);
+    }
+    return future;
+  }
+
+  private void executeWithRetry(
+      BoundQuery query,
+      Parameters executeParameters,
+      long queryStartNanos,
+      CompletableFuture<ResultSet> future) {
+    doExecute(
+        query,
+        executeParameters,
+        queryStartNanos,
+        future,
+        ex -> {
+          if (ex instanceof PreparedQueryNotFoundException) {
+            // This could happen due to a schema change between the statement preparation and now,
+            // as some schema change can invalidate preparation.
+            logger.debug(
+                "Prepared statement (id={}) was invalid when executed. This can happen due to a "
+                    + "conflicting schema change. Will re-prepare and retry.",
+                ((PreparedQueryNotFoundException) ex).id);
+            prepareAndRetry(query.bounded(), executeParameters, queryStartNanos, future);
+          } else {
+            future.completeExceptionally(ex);
+          }
+        });
+  }
+
+  private void doExecute(
+      BoundQuery query,
+      Parameters executeParameters,
+      long queryStartNanos,
+      CompletableFuture<ResultSet> successFuture,
+      Consumer<Throwable> onException) {
+    Statement statement = toPersistenceStatement(query);
+    connection
+        .execute(statement, executeParameters, queryStartNanos)
+        .thenAccept(
+            r ->
+                successFuture.complete(
+                    PersistenceBackedResultSet.create(connection, r, statement, executeParameters)))
+        .exceptionally(
+            ex -> {
+              onException.accept(ex);
```

Comment:
Not certain what is the the full suggestion.  `doExecute` uses a `Consumer` because `executeWithRetry` needs to add logic to the exception handling, logic that may execute asynchronously and not return an exception, so `completeExceptionally` doesn't work as-is.

But I'm sure those methods can be rewritten to use `CompletableFuture` method chaining, with some `#handle` and `#thenCompose`, but I find the current core easier to follow personally. This could very well be a case of personal preference and style.